lib/ex_aws/s3/download/sink.ex

```diff
@@ -3,17 +3,42 @@ alias Experimental.GenStage
 defmodule ExAws.S3.Download.Sink do
   use GenStage
 
-  def start_link(path) do
-    GenStage.start_link(__MODULE__, path, [])
+  defstruct [
+    :file_size,
+    :file,
+    bytes_written: 0
+  ]
+
+  def start_link(path, file_size) do
+    GenStage.start_link(__MODULE__, {path, file_size}, [])
+  end
+
+  def init({path, file_size}) do
+    state = %__MODULE__{
+      file: File.open!(path, [:write, :raw, :delayed_write, :binary]),
+      file_size: file_size,
+    }
+    {:consumer, state}
   end
 
-  def init(path) do
-    {:consumer, File.open!(path, [:write, :raw, :delayed_write, :binary])}
+  def handle_events(chunks, _from, state) do
+    :ok = :file.pwrite(state.file, chunks)
+
+    state = %{state | bytes_written: state.bytes_written + total_bytes(chunks)}
+
+    if done?(state) do
+      {:stop, :normal, state}
+    else
+      {:noreply, [], state}
+    end
```

Comment:
@benwilson512 the issue is that the producer won't terminate if the consumers exit. If you are exiting with normal reason, no link will break. So you will either need to have the producers check when all consumers are dead (handle_cancel will suffice) or use a non-normal exit or call GenStage.stop.